sdks/java/extensions/protobuf/src/main/java/org/apache/beam/sdk/extensions/protobuf/ProtoMessageSchema.java

```diff
@@ -115,6 +122,76 @@ public SchemaUserTypeCreator schemaTypeCreator(Class<?> targetClass, Schema sche
     return creator;
   }
 
+  public static <T extends Message> SimpleFunction<byte[], Row> getProtoBytesToRowFn(
+      Class<T> clazz) {
+    return new ProtoBytesToRowFn<>(clazz);
+  }
+
+  public static class ProtoBytesToRowFn<T extends Message> extends SimpleFunction<byte[], Row> {
+    private final ProtoCoder<T> protoCoder;
+    private final SerializableFunction<T, Row> toRowFunction;
+
+    public ProtoBytesToRowFn(Class<T> clazz) {
+      this.protoCoder = ProtoCoder.of(clazz);
+      this.toRowFunction = new ProtoMessageSchema().toRowFunction(TypeDescriptor.of(clazz));
+    }
+
+    @Override
+    public Row apply(byte[] bytes) {
+      try {
+        InputStream inputStream = new ByteArrayInputStream(bytes);
+        T message = protoCoder.decode(inputStream);
+        return toRowFunction.apply(message);
+      } catch (IOException e) {
+        throw new IllegalArgumentException("Could not decode row from proto payload.", e);
+      }
+    }
+  }
+
+  public static <T extends Message> SimpleFunction<Row, byte[]> getRowToProtoBytesFn(
+      Class<T> clazz) {
+    return new RowToProtoBytesFn<>(clazz);
+  }
+
+  public static class RowToProtoBytesFn<T extends Message> extends SimpleFunction<Row, byte[]> {
```

Comment:
If possible could you make these private and just have the `get*Fn` methods as the public API?

Reply:
Done.

Reply:
The `get*Fn` wants to receive Class<T extends Message> which cannot be provided from BeamKafkaProtoTable as it's not allowed to import from `com.google.protobuf`. I'll move the suppresion to ProtoMessageSchema and add a validation if the provided class extends Message.
If you know a cleaner solution then I'll gladly change it.

Reply:
Thanks! I think this approach is best for now. In the future I think we should try to isolate all the format specific logic in a module for just that format, and just have general purpose logic in SQL. This is closer to that vision imo.